src/controllers/estimation/estimation.ts

```diff
@@ -0,0 +1,267 @@
+      this.#outsideControllerNoUpdateStatuses
+    )
+    const estimation = await getEstimation(
+      baseAcc,
+      accountState,
+      op,
+      network,
+      this.#providers.providers[op.chainId.toString()],
+      feeTokens,
+      nativeToCheck,
+      bundlerSwitcher,
+      (e: ErrorRef) => {
+        if (!this) return
+        this.estimationRetryError = e
+        this.emitUpdate()
+      }
+    ).catch((e) => e)
+
+    const isSuccess = !(estimation instanceof Error)
+    if (isSuccess) {
+      this.estimation = getEstimationSummary(estimation)
+      this.error = null
+      this.status = EstimationStatus.Success
+      this.estimationRetryError = null
+      this.availableFeeOptions = this.#getAvailableFeeOptions(baseAcc)
+    } else {
+      this.estimation = null
+      this.error = estimation
+      this.status = EstimationStatus.Error
+      this.availableFeeOptions = []
+    }
+
+    // estimation.flags.hasNonceDiscrepancy is a signal from the estimation
+    // that the account state is not the latest and needs to be updated
+    if (
+      this.estimation &&
+      (this.estimation.flags.hasNonceDiscrepancy || this.estimation.flags.has4337NonceDiscrepancy)
+    )
+      // silenly continuing on error here as the flags are more like app helpers
+      this.#accounts.updateAccountState(op.accountAddr, 'pending', [op.chainId]).catch((e) => e)
+
+    this.hasEstimated = true
+    this.emitUpdate()
+  }
+
+  // it's initialized if it has estimated at least once
+  isInitialized() {
+    return this.hasEstimated
+  }
+
+  // has it estimated at least once without a failure
+  isLoadingOrFailed() {
+    return !this.hasEstimated || !!this.error
+  }
+
+  calculateWarnings() {
+    const warnings: Warning[] = []
+
+    if (this.estimationRetryError && this.status === EstimationStatus.Success) {
+      warnings.push({
+        id: 'estimation-retry',
+        title: this.estimationRetryError.message,
+        text: 'You can try to broadcast this transaction with the last successful estimation or wait for a new one. Retrying...',
+        promptBeforeSign: false
+      })
+    }
+
+    return warnings
+  }
+
+  get errors(): SignAccountOpError[] {
+    const errors: SignAccountOpError[] = []
+
+    if (this.isLoadingOrFailed() && this.estimationRetryError) {
+      // If there is a successful estimation we should show this as a warning
+      // as the user can use the old estimation to broadcast
+      errors.push({
+        title: `${this.estimationRetryError.message} ${
+          this.error
+            ? 'We will continue retrying, but please check your internet connection.'
+            : 'Automatically retrying in a few seconds. Please wait...'
+        }`,
+        code: 'ESTIMATION_ERROR'
+      })
+
+      return errors
+    }
+
+    if (!this.isInitialized()) return []
+
+    if (this.error) {
+      errors.push({
+        title: this.error.message,
+        code: 'ESTIMATION_ERROR'
```

Comment:
The code should be `typeof cause === 'string' && cause.length ? cause : 'ESTIMATION_ERROR'`